### PR TITLE
[Jest] Migrate `map.test.js`

### PR DESCRIPTION
**Pull request by wipfli**
- Move and rename
- Comment out `t.error()`, it breaks jscodeshift
- Run jscodeshift
- Run custom codemods and lint
- Manual edits

## Launch Checklist

Fixed TypeScript errors on 560 out of 2097 lines. Super happy to finally have reached this file :)

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by wipfli** [comment]
This is ready ,@HarelM, for review.

File: src/ui/map.test.ts [hunk 1]
```diff
@@ -16,22 +20,27 @@ function createStyleSource() {
...
+    setMatchMedia();
+});
 
-    t.afterEach((callback) => {
-        window.clearFakeXMLHttpRequest();
-        callback();
-    });
+describe('Map', () => {
+
+    jest.spyOn(global, 'XMLHttpRequest').mockReturnValue({
```
Comment (HarelM) [inline src/ui/map.test.ts:34]: You have sinon for that, it's better to use it as it has all the required mocking.
I think I converted a test that uses this, I don't remember right now which one.

File: src/ui/map.test.ts [hunk 2]
```diff
@@ -117,48 +118,44 @@ describe('Map', done => {
...
         setTimeout(() => {
-            map.off('load', fail);
+            map.off('load', done.fail);
             map.on('load', pass);
             map.setStyle(createStyle());
         }, 1);
 
-        function fail() { expect(false).toBeTruthy(); }
-        function pass() { t.end(); }
+        function pass() { done(); }
```
Comment (HarelM) [inline src/ui/map.test.ts:144]: Can't we simply call done instead of creating this method?

File: src/ui/map.test.ts [hunk 3]
```diff
@@ -271,50 +265,42 @@ describe('Map', done => {
...
 
         test('Map#isSourceLoaded', done => {
             const style = createStyle();
-            const map = createMap(t, {style});
+            const map = createMap({style});
 
             map.on('load', () => {
                 map.on('data', (e) => {
-                    if (e.dataType === 'source' && e.sourceDataType === 'metadata') {
+                    if (e.dataType === 'source' && (e as any).sourceDataType === 'metadata') {
```
Comment (HarelM) [inline src/ui/map.test.ts:303]: Cast to `MapDataEvent` or something similar (this type was introduced recently)

File: src/ui/map.test.ts [hunk 4]
```diff
@@ -337,25 +323,24 @@ describe('Map', done => {
...
         test('Map#areTilesLoaded', done => {
             const style = createStyle();
-            const map = createMap(t, {style});
+            const map = createMap({style});
             expect(map.areTilesLoaded()).toBe(true);
             map.on('load', () => {
                 const fakeTileId = new OverscaledTileID(0, 0, 0, 0, 0);
                 map.addSource('geojson', createStyleSource());
-                map.style.sourceCaches.geojson._tiles[fakeTileId.key] = new Tile(fakeTileId);
+                map.style.sourceCaches.geojson._tiles[fakeTileId.key] = new Tile(fakeTileId, undefined);
```
Comment (HarelM) [inline src/ui/map.test.ts:331]: Probably not the best way to test this as it uses internal members, but this wasn't introduced here so I don't expect it to be fixed, unless you have the time and will to do so :-)

File: src/ui/map.test.ts [hunk 5]
```diff
@@ -1040,12 +970,11 @@ describe('Map', done => {
         map.addControl(control);
         map.removeControl(control);
         expect(map._controls).toHaveLength(1);
```
Comment (HarelM) [inline src/ui/map.test.ts:972]: again, not the best condition as it uses internal fields...

File: src/ui/map.test.ts [hunk 6]
```diff
@@ -654,41 +626,37 @@ describe('Map', done => {
...
-            done();
         });
 
-        test('throws on invalid bounds', done => {
-            const map = createMap(t, {zoom:0});
+        test('throws on invalid bounds', () => {
+            const map = createMap({zoom:0});
             expect(() => {
-                map.setMaxBounds([-130.4297, 50.0642], [-61.52344, 24.20688]);
+                (map as any).setMaxBounds([-130.4297, 50.0642], [-61.52344, 24.20688]);
```
Comment (HarelM) [inline src/ui/map.test.ts:655]: is this `as any` really needed? I see that in the previous test you didn't use as any... same for two lines below.

File: src/ui/map.test.ts [hunk 7]
```diff
@@ -1947,23 +1862,23 @@ describe('Map', done => {
...
+        const map = createMap({style});
         map.on('render', () => {
             if (timer) clearTimeout(timer);
             timer = setTimeout(() => {
-                map.off('render');
-                map.on('render', t.fail);
-                expect(map._frameId).toBeFalsy();
+                map.off('render', undefined);
+                map.on('render', done.fail);
+                expect((map as any)._frameId).toBeFalsy();
```
Comment (HarelM) [inline src/ui/map.test.ts:1871]: Same here - `_frameId`...

**Review by HarelM — CHANGES_REQUESTED** [review]
Amazing work! well done! some minor comments...

File: src/ui/map.test.ts (hunk 6 above)
Comment (wipfli) [reply src/ui/map.test.ts:655]: It is. TypeScript prevents invalid arguments, which is what is being tested here.

Comment (HarelM) [reply src/ui/map.test.ts:655]: Didn't we agreed to remove tests of invalid input?

File: src/ui/map.test.ts (hunk 2 above)
Comment (wipfli) [reply src/ui/map.test.ts:144]: We can't. `map.on()` invokes the callback with an argument. 

https://github.com/maplibre/maplibre-gl-js/blob/e1c94dccdd7f52c90e2c3ec5de3bd48fcb75ab9f/src/util/evented.ts#L117-L119

Invoking `done()` with an argument fails the test.

Comment (HarelM) [reply src/ui/map.test.ts:144]: Use lambda then: `() => done()`

File: src/ui/map.test.ts (hunk 4 above)
Comment (wipfli) [reply src/ui/map.test.ts:331]: Leaving private members for another time.

File: src/ui/map.test.ts (hunk 5 above)
Comment (wipfli) [reply src/ui/map.test.ts:972]: Leaving private members for another time.

File: src/ui/map.test.ts (hunk 7 above)
Comment (wipfli) [reply src/ui/map.test.ts:1871]: Leaving private members for another time.

**Comment by wipfli** [comment]
Off-topic: Don't you find it also strange how the source code documentation works? For example, function parameters have to be declared optional both in the docs comments and in the actual function signature. Isn't there a better way to do this?

**Comment by wipfli** [comment]
Docs and code are out of sync for example here:

https://github.com/maplibre/maplibre-gl-js/blob/ea7db4d3942cf17ac65b91f7339ed5e22b9e3afb/src/ui/events.ts#L288-L311

**Comment by wipfli** [comment]
Thanks for the review. I'm really happy that you always make these excellent code reviews. I think I addressed all comments. Please close conversations where you are OK with the answer.

File: src/ui/events.ts [hunk 8]
```diff
@@ -7,6 +7,7 @@ import {extend} from '../util/util';
 import type Map from './map';
 import type LngLat from '../geo/lng_lat';
 import {SourceSpecification} from '../style-spec/types';
+import {Coordinates} from '../source/image_source';
```
Comment (HarelM) [inline src/ui/events.ts:10]: Import type

File: src/ui/events.ts [hunk 9]
```diff
@@ -308,6 +309,11 @@ export type MapLibreZoomEvent = {
 export type MapDataEvent = {
   type: string;
   dataType: string;
+  isSourceLoaded?: boolean;
+  source?: Object;
```
Comment (HarelM) [inline src/ui/events.ts:313]: These members are part of mapsourcedataevent I believe...?

**Comment by HarelM** [comment]
Yea, I agree that docs sync is annoying.
Maybe the type should not be part of the docs just the explanation of the parameter? IDK...

File: src/ui/events.ts (hunk 9 above)
Comment (wipfli) [reply src/ui/events.ts:313]: Yes. You added `MapSourceDataEvent` when [fixing the typings generated by TypeScript](https://github.com/maplibre/maplibre-gl-js/commit/ddf69421c6ae34c808afefec309a5beecdb7500e#diff-b8d6a5f83861b40661fd2ac4ae7a12b899d83ba7d4cdc522288936db93e743b0). I don't know what to do now. Should I update the docs comment or change the code?

Comment (HarelM) [reply src/ui/events.ts:313]: Ahh, I now see the mismatch. I had a discussion somewhere (related to geojson events data) in one of the PRs about exactly this.
I'm not sure which is the right one. As far as I understand, but I might be wrong, these parameters are not optional in the data source event and there isn't an event that fires something with just type and datatype. But I'm not 100% sure. Feel free to investigate :-)

Comment (HarelM) [reply src/ui/events.ts:313]: See the following PR: #695 - last resolved conversation is talking about this more or less.
@vanilla-lake can you add your input here too as well, I think you know this better than I do.

**Comment by wipfli** [comment]
I rebased onto main to include the new tests from #727. Commits you might want to look at @HarelM are the last three:

https://github.com/maplibre/maplibre-gl-js/pull/723/commits/31f16e357c103de448f0c4e00efb8894544227ce
https://github.com/maplibre/maplibre-gl-js/pull/723/commits/0cbf5139b0dd45078438945216c55d57e2520e75
https://github.com/maplibre/maplibre-gl-js/pull/723/commits/9e96a3808431bdaff843140ed180eeb82897f28b

**Comment by HarelM** [comment]
Reviewed, no new comments. The only item left to reconcile is the above conversation, let's wait for @vanilla-lake 's response.

File: src/ui/events.ts (hunk 9 above)
Comment (vanilla-lake) [reply src/ui/events.ts:313]: As far as I can tell, @HarelM is right: those properties will only get a value in the case of source data events.

Comment (wipfli) [reply src/ui/events.ts]: Thanks @vanilla-lake

File: src/ui/events.ts [hunk 10]
```diff
@@ -308,6 +308,7 @@ export type MapLibreZoomEvent = {
 export type MapDataEvent = {
   type: string;
   dataType: string;
+  sourceDataType: string;
```
Comment (HarelM) [inline src/ui/events.ts:311]: Should be of type `MapSourceDataType`?

Comment (wipfli) [reply src/ui/events.ts:311]: Correct, thanks!

**Review by HarelM — APPROVED** [review]

**Pull request merged**